substra/sdk/backends/local/backend.py

```diff
@@ -89,6 +97,49 @@ def _is_local(key: str):
         """
         return key.startswith("local_")
 
+    def __copy_file(self,
+                    file_path: typing.Union[str, pathlib.Path],
+                    key: str,
+                    exist_ok: bool):
+        """Copy file or directory into the local temp dir to mimick
+        the remote backend that saves the files given by the user.
+
+        If exist_ok is True, no error is thrown if there is a file with the same
```

Comment:
Yes, it is not the cleanest way :/

If we save the file after adding it to the db then we still have to get the information that the file already existed and not to change it.

So I changed the code and as you suggested I now check if the object exists in the db as a first step when adding an asset
--> the `exist_ok` parameter is now a parameter of the `check_exist` function and not the `add` function.

Reply:
Thanks, that's better.

I'm not 100% convinced with the proposed `check_exists()` function: it's hard to understand by reading its signature what it does and what it returns. IMO it should be the backend responsibility to handle this. We could maybe move out the `exist_ok` option?

```python
try:
    asset = self._db.get(type_=schemas.Type.DataSample, key=key)
except NotFoundError:
    pass
else:
    if exist_ok:
        return asset
    raise AlreadyExistError     
```

Or maybe:
```python
    asset = self._db.check_exists(type_=schemas.Type.DataSample, key=key)
    if asset:
        if exist_ok:
            return asset
        raise AlreadyExistError    
```

Reply:
I am OK with moving it to the backend, I am less sure about how to make the function clearer. My problem is code duplication, I already have a bit of a problem with copying
```python
asset = self._db.check_exist(type_=schemas.Type.Dataset, key=key, exist_ok=exist_ok)
if asset:
    return asset
```
in each `_add_...` function.

What do you think about creating `_compute_key_...` functions for each asset, and, in the `add` function, have this:
```python
def add(self, spec, exist_ok, spec_options=None):
  method_name = f"_compute_key_{spec.__class__.type_.value}"
  compute_key = getattr(self, method_name)
  key = method_name(spec)
  try:
      asset = self._db.get(type_=spec.__class__.type_.value, key=key)
  except NotFoundError:
      pass
  else:
      if exist_ok:
          return asset
      raise AlreadyExistError

.... # call the _add_{asset_type} function
``` 
There might be an additional modification to do for the `spec.is_many` case though, but at least we would have less code duplication on this and we would not need to hardcode the `type_` in the `self._db.get` function.

Reply:
As we talked about, I put the `compute_key` fct in `schemas.py`, that required to move the `fs` and `hasher` module from sdk/backends/local to sdk so that sdk/schemas does not call backends/local...

Compute plan is a bit of a special case as we may create one when creating a traintuple, so I decided to make the `compute_key` fct static rather than create a ComputePlanSpec only to compute the key.